lib/OpenQA/WebAPI/Controller/API/V1/Job.pm

```diff
@@ -219,60 +220,174 @@ sub overview {
 
 =over 4
 
-=item create()
+=item _eval_param_grouping()
 
-Creates a job given a list of settings passed as parameters. TEST setting/parameter
-is mandatory and should be the name of the test.
+Removes job-specific parameters from the specified hash and returns a new hash
+containing the removed parameters (where the keys are job suffixes).
+
+So "{foo => 1, bar:jobA => 2, baz:jobB => 3}" will be turned into "{foo => 1}"
+returnig "{jobA => {bar => 2}, jobB => {baz => 3}}".
 
 =back
 
 =cut
 
-sub create {
-    my $self = shift;
-    my $params = $self->req->params->to_hash;
-    my $is_clone_job = delete $params->{is_clone_job} // 0;
+sub _eval_param_grouping ($params) {
+    my %grouped_params;
+    for my $param_key (keys %$params) {
+        my $job_suffix_start = rindex($param_key, ':');
+        next unless $job_suffix_start >= 0;
+        my $job_suffix = substr $param_key, $job_suffix_start + 1;
+        my $setting_key = substr $param_key, 0, $job_suffix_start;
+        $grouped_params{$job_suffix}->{$setting_key} = delete $params->{$param_key};
+    }
+    return \%grouped_params;
+}
+
+=over 4
+
+=item _eval_dependency()
+
+Removes the specified $settings_keys from $job_settings populating $self->{_dependencies}
+with the dependency information from the removed hash entry value.
+
+=back
+
+=cut
 
+sub _eval_dependency ($self, $child_job_suffix, $job_settings, $setting_key, $type) {
+    return undef unless $child_job_suffix;
+    my $job_suffixes = delete $job_settings->{$setting_key};
+    return undef unless defined $job_suffixes;
+    my $deps = $self->{_dependencies};
+    for my $parent_job_suffix (split /\s*,\s*/, $job_suffixes) {
+        push @$deps,
+          {
+            child_job_id => $child_job_suffix,
+            parent_job_id => $parent_job_suffix,
+            dependency => $type,
+          };
+    }
```

Comment:
```suggestion
    my $deps = ($self->{_dependencies}, map { {
            child_job_id => $child_job_suffix,
            parent_job_id => $_,
            dependency => $type } } split /\s*,\s*/, $job_suffixes;
```

Reply:
This makes no sense. `$self->{_dependencies}` is supposed to be populated here.

Reply:
oh, I see. I misread `my $deps = $self->{_dependencies} ` as deep copy which it isn't. So maybe

```suggestion
$self->{_dependencies} = ($self->{_dependencies}, map { {
            child_job_id => $child_job_suffix,
            parent_job_id => $_,
            dependency => $type } } split /\s*,\s*/, $job_suffixes);
```

because you can merge a second hash (here created by "map") into the first like `%h = (%h, %j)` or here more like `%h = (%h, map { … } @j)`

Reply:
That's harder to read and unlikely to be more efficient. And it gets even uglier because you needed to make a new hash-ref (by wrapping in `[]` instead of `()`) and you needed to de-ref `$self->{_dependencies}` to retain the flat structure.

Reply:
I didn't invent it. I thought this is the way how to merge hashes in perl

Reply:
I think of it more as "concatenation". "Merge" means to me something becomes part of something else but here we make something new and throw away the original parts.

Reply:
Yes, I meant that `%h = (%h, %j)` is the way to concatenate hashes in perl